tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,292 @@
+"""Tests for the iCloud config flow."""
+from unittest.mock import patch, Mock, MagicMock
+import pytest
+
+from pyicloud.exceptions import PyiCloudFailedLoginException
+
+from homeassistant import data_entry_flow
+from homeassistant.components.icloud import config_flow
+
+from homeassistant.components.icloud.config_flow import (
+    CONF_TRUSTED_DEVICE,
+    CONF_VERIFICATION_CODE,
+)
+from homeassistant.components.icloud.const import (
+    DOMAIN,
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DEFAULT_MAX_INTERVAL,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.helpers.typing import HomeAssistantType
+
+from tests.common import MockConfigEntry
+
+USERNAME = "username@me.com"
+PASSWORD = "password"
+ACCOUNT_NAME = "Account name 1 2 3"
+ACCOUNT_NAME_FROM_USERNAME = None
+MAX_INTERVAL = 15
+GPS_ACCURACY_THRESHOLD = 250
+
+TRUSTED_DEVICES = [
+    {"deviceType": "SMS", "areaCode": "", "phoneNumber": "*******58", "deviceId": "1"}
+]
+
+
+@pytest.fixture(name="service")
+def mock_controller_service():
+    """Mock a successful service."""
+    with patch(
+        "homeassistant.components.icloud.config_flow.PyiCloudService"
+    ) as service_mock:
+        service_mock.authenticate.return_value = Mock()
+        service_mock.return_value.requires_2fa = True
+        yield service_mock
+
+
+@pytest.fixture(name="service_with_cookie")
+def mock_controller_service_with_cookie():
+    """Mock a successful service while already authenticate."""
+    with patch(
+        "homeassistant.components.icloud.config_flow.PyiCloudService"
+    ) as service_mock:
+        service_mock.authenticate.return_value = Mock()
+        service_mock.return_value.requires_2fa = False
+        yield service_mock
+
+
+@pytest.fixture(name="service_send_verification_code_failed")
+def mock_controller_service_send_verification_code_failed():
+    """Mock a failed service during sending verification code step."""
+    with patch(
+        "homeassistant.components.icloud.config_flow.PyiCloudService"
+    ) as service_mock:
+        service_mock.authenticate.return_value = Mock()
+        service_mock.return_value.requires_2fa = False
+        service_mock.return_value.trusted_devices = TRUSTED_DEVICES
+        service_mock.send_verification_code = Mock(return_value=False)
```

Comment:
We need to insert `.return_value` between `service_mock` and `.send_verification_code`.

Reply:
Does not work with :

<img width="904" alt="Capture d’écran 2019-12-03 à 15 55 42" src="https://user-images.githubusercontent.com/14821482/70061917-65553b00-15e5-11ea-984c-194c4fdc4945.png">

Reply:
Tried with : 
```python
service_mock.return_value.send_verification_code = Mock(return_value=False)
```
```python
service_mock.return_value.send_verification_code.return_value = Mock(return_value=False)
```
```python
service_mock.return_value.send_verification_code.return_value = MagicMock(return_value=False)
```
```python
service_mock.return_value.send_verification_code.return_value = False
```
```python
service_mock.return_value.send_verification_code = False
```

Reply:
What do you think @MartinHjelmare ?

I’m tired of changing my working tests 😅

Reply:
So what do you think ?

Reply:
See below.